src/caffe/parallel.cpp

```diff
@@ -391,15 +392,14 @@ void P2PSync<Dtype>::run(const vector<int>& gpus) {
   LOG(INFO)<< "GPUs pairs " << s.str();
 
   SolverParameter param(solver_->param());
-  vector<shared_ptr<P2PSync<Dtype> > > syncs(gpus.size());
 
   // Build the GPU tree by finding the parent for each solver
   for (int attempts = 0; attempts < pairs.size(); ++attempts) {
     for (int i = 1; i < pairs.size(); ++i) {
-      if (!syncs[i].get()) {
+      if (!syncs->at(i).get()) {
```

Comment:
Thanks for reiterating the difference between [i] and at(i). It is clear.

The output arguments in Prepare() is however different from what's described in https://github.com/BVLC/caffe/pull/945. Prepare() in this patch corresponds to the original code here: https://github.com/BVLC/caffe/blob/master/src/caffe/parallel.cpp#L394-L412

syncs starts with a vector of empty shared_ptrs (i.e. null). Inside Prepare(), those shared pointers gets assigned. In https://github.com/BVLC/caffe/pull/945, vector<Blob<Dtype>* > is already initialized (i.e. the pointers inside the vector points valid objects, not null), those Forward or Backward functions only change the content of the blobs, but not replace or delete the blobs.

I had tried const reference in the first place, but had issues with this reset function. 
https://github.com/BVLC/caffe/pull/3588/files#diff-55c214d759a1a3d26cfdaf807eb86b10L412
Hence it is now a pointer to the vector.